services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_compose_specs.py

```diff
@@ -95,26 +99,54 @@ def _inject_paths_mappings(
         service_content["environment"] = _environment_section.export_as_list(env_vars)
 
 
-def _replace_env_vars_in_compose_spec(
-    stringified_service_spec: str, resolved_registry_url: str, service_tag: str
-) -> str:
-    stringified_service_spec = stringified_service_spec.replace(
-        MATCH_SIMCORE_REGISTRY, resolved_registry_url
-    )
-    stringified_service_spec = stringified_service_spec.replace(
-        MATCH_SERVICE_VERSION, service_tag
-    )
-    return stringified_service_spec
+def _update_resource_limits_and_reservations(
+    service_resources: ServiceResources, service_spec: ComposeSpecLabel
+) -> None:
+    # example: '2.3' -> 2 ; '3.7' -> 3
+    docker_compose_major_version: int = int(service_spec["version"].split(".")[0])
+
+    for spec_service_key, spec in service_spec["services"].items():
+        resources: Resources = service_resources[spec_service_key].resources
+        logger.debug("Resources for %s: %s", spec_service_key, f"{resources=}")
+
+        cpu: ResourceValue = resources["CPU"]
+        memory: ResourceValue = resources["RAM"]
+
+        if docker_compose_major_version >= 3:
+            # compos spec version 3 and beyond
+            deploy = spec.get("deploy", {})
+            resources = deploy.get("resources", {})
+            limits = resources.get("limits", {})
+            reservations = resources.get("reservations", {})
+
+            # assign limits
+            limits["cpus"] = float(cpu.limit)
+            limits["memory"] = f"{memory.limit}"
+            # assing reservations
+            reservations["cpus"] = float(cpu.reservation)
+            reservations["memory"] = f"{memory.reservation}"
+
+            resources["reservations"] = reservations
+            resources["limits"] = limits
+            deploy["resources"] = resources
+            spec["deploy"] = deploy
+        else:
+            # compos spec version 2
+            spec["mem_limit"] = f"{memory.limit}"
+            spec["mem_reservation"] = f"{memory.reservation}"
+            # NOTE: there is no distinction between limit and reservation, taking the higher value
+            spec["cpus"] = float(max(cpu.limit, cpu.reservation))
 
 
-def assemble_spec(
+async def assemble_spec(
```

Comment:
Why is this async?

Reply:
some leftover, it was removed

Reply:
But you have added it! So is there and async or not?

IMO there is NOT and if so ... make sure that ``assemble_spec`` is not awaited (as it is now!)